### PR TITLE
APPT-XXX: Disable nullable reference types and fix warnings

### PR DESCRIPTION
**Pull request by sambiramairelogic**
Disables nullable reference types in every C# project and resolves all currently outstanding warnings.

### CODE REVIEW

File: src/api/Nhs.Appointments.Api/Functions/ApplyAvailabilityTemplateFunction.cs [hunk 1]
```diff
@@ -17,6 +17,8 @@ namespace Nhs.Appointments.Api.Functions;
 public class ApplyAvailabilityTemplateFunction(IAvailabilityService availabilityService, IValidator<ApplyAvailabilityTemplateRequest> validator, IUserContextProvider userContextProvider, ILogger<ApplyAvailabilityTemplateFunction> logger, IMetricsRecorder metricsRecorder) 
     : BaseApiFunction<ApplyAvailabilityTemplateRequest, EmptyResponse>(validator, userContextProvider, logger, metricsRecorder)
 {
+    private readonly IUserContextProvider _userContextProvider = userContextProvider;
```
Comment (vicr1) [inline src/api/Nhs.Appointments.Api/Functions/ApplyAvailabilityTemplateFunction.cs:20]: Not sure about this change

File: src/api/Nhs.Appointments.Api/Functions/RemoveUserFunction.cs [hunk 2]
```diff
@@ -21,6 +21,8 @@ public class RemoveUserFunction(
     IMetricsRecorder metricsRecorder)
     : BaseApiFunction<RemoveUserRequest, RemoveUserResponse>(validator, userContextProvider, logger, metricsRecorder)
 {
+    private readonly IUserContextProvider _userContextProvider = userContextProvider;
```
Comment (vicr1) [inline src/api/Nhs.Appointments.Api/Functions/RemoveUserFunction.cs:24]: Are these required?

File: src/api/Nhs.Appointments.Api/Notifications/ConsoleLogNotifications.cs [hunk 3]
```diff
@@ -100,7 +100,7 @@ public void AddConsumeTask(Task task)
             throw new NotImplementedException();
         }
 
-        public T AddOrUpdatePayload<T>(PayloadFactory<T> addFactory, UpdatePayloadFactory<T> updateFactory) where T : class
+        public TT AddOrUpdatePayload<TT>(PayloadFactory<TT> addFactory, UpdatePayloadFactory<TT> updateFactory) where TT : class
```
Comment (vicr1) [inline src/api/Nhs.Appointments.Api/Notifications/ConsoleLogNotifications.cs:103]: TPayload would be a good name

File: tests/Nhs.Appointments.Api.Integration/Scenarios/CreateAvailability/GetAvailabilityCreatedEventsFeatureSteps.cs [hunk 4]
```diff
@@ -12,9 +12,9 @@ namespace Nhs.Appointments.Api.Integration.Scenarios.CreateAvailability
     [FeatureFile("./Scenarios/CreateAvailability/GetAvailabilityCreatedEvents.feature")]
     public sealed class GetAvailabilityCreatedEventsFeatureSteps : BaseCreateAvailabilityFeatureSteps
     {
-        private HttpResponseMessage _response;
-        private HttpStatusCode _statusCode;
-        private IEnumerable<AvailabilityCreatedEvent> _actualResponse;
+        private new HttpResponseMessage _response;
```
Comment (vicr1) [inline tests/Nhs.Appointments.Api.Integration/Scenarios/CreateAvailability/GetAvailabilityCreatedEventsFeatureSteps.cs:15]: if these are hiding members perhaps they should just use the underlying members

**Review by vicr1 — CHANGES_REQUESTED** [review]

File: src/api/Nhs.Appointments.Api/Functions/RemoveUserFunction.cs (hunk 2 above)
Comment (sambiramairelogic) [reply src/api/Nhs.Appointments.Api/Functions/RemoveUserFunction.cs]: This is the warning they generate: 
image
I'm unsure how else to resolve this warning as I'll be honest I don't 100% understand what it means. I'll do some googling and get my head round it, but for expediency I've reverted these changes. Shall I leave 4 warnings unresolved and tackle in a future PR?

File: tests/Nhs.Appointments.Api.Integration/Scenarios/CreateAvailability/GetAvailabilityCreatedEventsFeatureSteps.cs (hunk 4 above)
Comment (sambiramairelogic) [reply tests/Nhs.Appointments.Api.Integration/Scenarios/CreateAvailability/GetAvailabilityCreatedEventsFeatureSteps.cs]: Yeah it turns out it wasn't even needed on the base class, so I've removed it from there and the new keyword from here

**Review by vicr1 — APPROVED** [review]

**Review by vicr1 — APPROVED** [review]

**Pull request merged**